library/alloc/src/raw_vec.rs

```diff
@@ -315,8 +315,24 @@ impl<T, A: Allocator> RawVec<T, A> {
     /// #   vector.push_all(&[1, 3, 5, 7, 9]);
     /// # }
     /// ```
+    #[inline]
     pub fn reserve(&mut self, len: usize, additional: usize) {
-        handle_reserve(self.try_reserve(len, additional));
+        // Callers expect this function to be very cheap when there is already sufficient capacity.
+        // Therefore, we move all the resizing and error-handling logic from grow_amortized and
+        // handle_reserve behind a call, while making sure that the this function is likely to be
+        // inlined as just a comparison and a call if the comparison fails.
+        #[cold]
```

Comment:
Why mark this as `cold` instead of `inline(never)`? I would expect this to be called somewhat frequently in a normal program.

Reply:
From what I have read in other comments cold simply means llvm's coldcc. So it minimizes the impact on the caller, isn't that the goal here? We assume the caller code to be hotter.

Reply:
In some places inside vec (and some other modules) it's annotated via
```rust
#[cold]
#[inline(never)]
```

Reply:
@jyn514 That's kind of right and I'm taking a new look at this PR with the same goal of reducing the amount of time spent in `reserve` calls that don't cause any capacity change.

It looks to me like most of the `Vec::reserve` calls are guarded by a capacity check (such as the one in `push`). But some aren't. This manifests in the `serde_json` benchmarks because serializers do a lot of `Write::write_all` calls which is implemented by `Vec::extend_from_slice`, which is implemented (eventually) by `append_elements`, which has an unguarded `self.reserve`. I'm assessing adding an `if self.len() + additional > self.capacity()` guard around the call in `append_elements`, and going to take a look at either guarding _all_ the `reserve` calls or sinking the check into `Vec::reserve`, and deduplicating the check around a lot of the call sites in `Vec`.